python-server/pvCamera.py

```diff
@@ -0,0 +1,249 @@
+    return True
+
+async def check_signal_connection(signal, name, websocket):
+    if not signal.connected:
+        print('array data pv not connected, exiting')
+        await websocket.send_text(json.dumps({'error': f"The {name} pv could not connect"}))
+        await websocket.close()
+        return False
+    return True
+
+def update_enum_lists(settingSignals, colorModeEnumList, dataTypeEnumList):
+    #ColorMode and DataType should have an enum_strs attribute containing an array of
+    #strings whose index corresponds to the value held by the pv
+    #Overwrite the defaults with the actual values, if values don't exist than accept default
+    colorModeEnumList = getattr(settingSignals['colorMode'], 'enum_strs', colorModeEnumList)
+    dataTypeEnumList = getattr(settingSignals['dataType'], 'enum_strs', dataTypeEnumList)
+    return colorModeEnumList, dataTypeEnumList
+
+def update_dimensions(settingSignals, buffer):
+    colorModeValue = settingSignals['colorMode'].get()
+    dataTypeValue = settingSignals['dataType'].get()
+    tempDimensions = {
+        'x': settingSignals['sizeX'].get() - settingSignals['startX'].get(),
+        'y': settingSignals['sizeY'].get() - settingSignals['startY'].get(),
+        'colorMode': colorModeEnumList[colorModeValue],
+        'dataType': dataTypeEnumList[dataTypeValue]
+    }
+    buffer.put_nowait((None, time.time(), tempDimensions))
+
+def format_array(rawImageArray, height, width, colorMode, dataType):
+    array_data = np.array(rawImageArray, dtype=dtype_map[dataType])
+    array_data = normalize_array_data(array_data, dataType)
+    array_data, mode = reshape_array(array_data, height, width, colorMode)
+    return array_data, mode
+
+# Main loop for streaming images
+
+async def handle_streaming(websocket, buffer):
+    try:
+        while True:
+            rawImageArray, timestamp, updatedSettings = await buffer.get()
+
+            if updatedSettings:
+                currentSettings = updatedSettings
+                await websocket.send_text(json.dumps(currentSettings))
+                continue
+
+            height, width, colorMode, dataType = currentSettings['y'], currentSettings['x'], currentSettings['colorMode'], currentSettings['dataType']
+
+            try:
+                #array_data = np.array(rawImageArray, dtype=dtype_map[dataType])
+                #array_data = normalize_array_data(array_data, dataType)
+                #array_data, mode = reshape_array(array_data, height, width, colorMode)
+                #await send_image(array_data, websocket, mode)
+
+                #change to make non blocking, PR#13
+                format_array_coroutine = asyncio.to_thread(format_array, rawImageArray, height, width, colorMode, dataType)
+                array_data, mode = await format_array_coroutine
+                
+                get_buffer_coroutine = asyncio.to_thread(send_image, array_data, websocket, mode)
+                buffered = await get_buffer_coroutine
```

Comment:
It's typical to await in the same line as defining the co-routine:

```

                #change to make non blocking, PR#13
                array_data, mode = await asyncio.to_thread(format_array, rawImageArray, height, width, colorMode, dataType)
                buffered = await asyncio.to_thread(send_image, array_data, websocket, mode)
           
```

Also, thinking more, let's combine these two separate threaded functions into one. Threads have a small cost, and reducing them is good when you can.

Reply:
-Put these two calls into new function `get_buffer()` line 217
-Called function same line as await
-Revised try catch blocks